### PR TITLE
Chunked storage response streaming

### PR DESCRIPTION
**Pull request by gwicke**
This patch changes the behavior of read requests to stream body chunks as they
become available, rather than collecting & concatenating all chunks before
returning a response.

This patch also bumps up the default chunk size to 1.8mb, in line with the
benchmark results from https://phabricator.wikimedia.org/T122028. This won't
provide special compression efficiency with deflate, but will provide a decent
combination of compression ratio and scalability using brotli compression.

### CODE REVIEW

File: sys/key_rev_large_value.js [hunk 1]
```diff
@@ -157,11 +158,6 @@ function returnRevision(req, metadata) {
     return function(dbResult) {
         if (dbResult.body && dbResult.body.items && dbResult.body.items.length) {
             var row = dbResult.body.items[0];
-            return {
```
Comment (d00rman) [inline sys/key_rev_large_value.js]: euh, so this function returns nothing any more?

File: sys/key_rev_large_value.js [hunk 2]
```diff
@@ -232,19 +212,18 @@ ChunkedBucket.prototype.getRevision = function(hyper, req) {
...
-            metadata.value = self._joinData(sections);
-            return {
-                status: 200,
-                body: {
-                    items: [ metadata ]
-                }
-            };
         })
-        .then(returnRevision(req, metadata));
+        .then(function() { return byteStream.end(); });
```
Comment (Pchelolo) [inline sys/key_rev_large_value.js]: We need to add a `catch` here to avoid unhandled exceptions

**Comment by Pchelolo** [comment]
Oh, already needs a rebase..

File: sys/key_rev_large_value.js [hunk 3]
```diff
@@ -218,13 +195,16 @@ ChunkedBucket.prototype._getMetadata = function(hyper, req) {
 ChunkedBucket.prototype.getRevision = function(hyper, req) {
     var rp = req.params;
     var self = this;
+    var chunksTable = self._chunksTableName(req);
+    var chunkURI = new URI([rp.domain, 'sys', 'table', chunksTable, '']);
     return this._getMetadata(hyper, req)
     .then(function(metadata) {
-        return P.all(range(metadata.num_chunks).map(function(chunkId) {
+        var byteStream = new stream.PassThrough();
+        P.each(range(metadata.num_chunks), function(chunkId) {
```
Comment (Pchelolo) [inline sys/key_rev_large_value.js]: A comment about `missing return on purpose` could be nice.

File: sys/key_rev_large_value.js [hunk 4]
```diff
@@ -232,19 +212,24 @@ ChunkedBucket.prototype.getRevision = function(hyper, req) {
...
-            };
         })
-        .then(returnRevision(req, metadata));
+        .then(function() { byteStream.end(); })
+        .catch(function(e) {
+            hyper.log('error/key_rev_large_value/body', e);
+            // FIXME: Figure out a way to abort the response, but don't signal
+            // success via .end().
+        });
+
```
Comment (d00rman) [inline sys/key_rev_large_value.js]: Nit: too many empty lines here.

File: sys/key_rev_large_value.js (hunk 3 above)
Comment (d00rman) [inline sys/key_rev_large_value.js]: +1

**Comment by d00rman** [comment]
As it currently stands, we can only ever return a 200. I'm not sure that's good, tbh. I'd prefer for us to find a way to communicate errors in a sane way before embarking on this *streamisation* of the code base.

**Comment by gwicke** [comment]
We can (and do) return errors if the metadata is not found (requested item does not exist). With streaming, we inherently can't signal a different HTTP response code if something goes wrong while streaming a body part. It would be nice, however, to be able to force-close the connection in a way that is guaranteed to be picked up as an error by the client, rather than letting the connection time out.

**Comment by d00rman** [comment]
If the metada is found, but the chunks aren't there, we'd return a 200 with an empty body here.

**Comment by gwicke** [comment]
> If the metada is found, but the chunks aren't there, we'd return a 200 with an empty body here.

More precisely, the client response would time out.

**Comment by gwicke** [comment]
Closing / archiving this, as we found that per-chunk overheads dominated at small chunk sizes. We could revive this if we had a need to store & stream really large chunks (such a media).

**Pull request closed without merging**